src/Raven.Server/Documents/Queries/Dynamic/DynamicQueryToIndexMatcher.cs

```diff
@@ -57,61 +59,103 @@ public Explanation(string index, string reason)
-            var definitions = _indexStore.GetIndexesForCollection(query.ForCollection)
-                .Where(x => x.Type.IsAuto() && (query.IsGroupBy ? x.Type.IsMapReduce() : x.Type.IsMap()))
-                .Select(x => x.Definition as AutoIndexDefinitionBase)
-                .ToList();
+            if (TryGetAutoIndexDefinitions(out var definitions) == false)
+                return DynamicQueryMatchResult.Failure;
 
-            if (definitions.Count == 0)
-                return new DynamicQueryMatchResult(string.Empty, DynamicQueryMatchType.Failure);
+            if (TryConvertToMatches(out var results) == false)
+                return DynamicQueryMatchResult.Failure;
 
-            var results = definitions.Select(definition => ConsiderUsageOfIndex(query, definition, explanations))
-                    .Where(result => result.MatchType != DynamicQueryMatchType.Failure)
-                    .GroupBy(x => x.MatchType)
-                    .ToDictionary(x => x.Key, x => x.ToArray());
-
-            if (results.TryGetValue(DynamicQueryMatchType.Complete, out DynamicQueryMatchResult[] matchResults) && matchResults.Length > 0)
+            if (results.TryGetValue(DynamicQueryMatchType.Complete, out var matchResults) && matchResults.Count > 0)
             {
                 return SelectIndexMatchingCompletely(explanations, matchResults);
             }
 
-            if (results.TryGetValue(DynamicQueryMatchType.CompleteButIdle, out matchResults) && matchResults.Length > 0)
+            if (results.TryGetValue(DynamicQueryMatchType.CompleteButIdle, out matchResults) && matchResults.Count > 0)
             {
                 return SelectIndexMatchingCompletely(explanations, matchResults);
             }
 
-            if (results.TryGetValue(DynamicQueryMatchType.Partial, out matchResults) && matchResults.Length > 0)
+            if (results.TryGetValue(DynamicQueryMatchType.Partial, out matchResults) && matchResults.Count > 0)
             {
                 return matchResults.OrderByDescending(x => x.NumberOfMappedFields).First();
             }
 
-            return new DynamicQueryMatchResult(string.Empty, DynamicQueryMatchType.Failure);
+            return DynamicQueryMatchResult.Failure;
+
+            bool TryGetAutoIndexDefinitions(out List<AutoIndexDefinitionBase> list)
+            {
+                list = null;
+
+                foreach (var index in _indexStore.GetIndexesForCollection(query.ForCollection))
+                {
+                    if ((index.Type.IsAuto() && (query.IsGroupBy ? index.Type.IsMapReduce() : index.Type.IsMap())) == false)
+                        continue;
+
+                    list ??= new List<AutoIndexDefinitionBase>();
+
+                    list.Add((AutoIndexDefinitionBase)index.Definition);
+                }
+
+                return list?.Count > 0;
+            }
+
+            bool TryConvertToMatches(out Dictionary<DynamicQueryMatchType, List<DynamicQueryMatchResult>> dictionary)
```

Comment:
Do we have to do it like this?
Can't we just use a local variable for storing the complete option (and if finding another one, switching if we have a better ranking)?

Reply:
Actually this isn't even close to main cost there. I think I can optimize it even further.

BTW. Here are the results:
![image](https://user-images.githubusercontent.com/446555/80303532-1df86280-87b1-11ea-959d-ae08ee19b981.png)

Reply:
The main cost is here:

![image](https://user-images.githubusercontent.com/446555/80303618-a1b24f00-87b1-11ea-8e3d-962637609572.png)

We are getting stats to check if index is invalid.

1. We probably do not need to take whole stats
2. GetMemoryStats is taking 77% of the time from which 58% are taken by GetDirectoryName!

Reply:
Still need to check where we actually want to retrieve that Memory stats, but:

![image](https://user-images.githubusercontent.com/446555/80304021-eb9c3480-87b3-11ea-88ed-3b2ae5936010.png)

We are down from 49300ms to 10500ms now.

Reply:
I have another way to do that, and I think it reduces a lot of the collection code